## Difference between traditional and current feature detection?
<img src="files/2.png">

### What is deep in Deep Learning?

As Jeff Dean quoted "deep" refers to the large number of layers.

## Image Fundamentals

### Pixels - The building blocks of image.
A pixel is considered as "color" or "intensity" of the light at the particular point. Image can be conceptualize as multidimensional matrix.
Most pixels are represented in two ways:
1. Grayscale/ Single channel -  Each pixel is a scalar value between 0 and 255, where "0" corresponds to "black" and "255" corresponds to "white". And values between 0 and 255 correspond to various shades of grey.
2. Color/ Multiple channel - Each pixel is demonstrated in the **RGB** color space and no longer holds a single value like grayscale. Each pixel is represented by **3 values**, each for red, green, and blue. Color in RGB channel is defined by **the amount of color**. Each red, blue, and green have values in range **[0, 255]**, where 0 indicates no representation and 255 full representation. The RGB color space is an example of an additive color space: the more of each color is added, the brighter the pixel becomes and closer to white.

We count from 0 rather than 1.<img src="files/3.png">

In [ ]:
import cv2
image = cv2.imread("files/2.png")
print(image.shape)
cv2.imshow("Image", image)
cv2.waitKey(0)

(411, 244, 3)


The notation is **(height, width, depth)**. This image has a width of 244 pixels (the number of columns), a height of 411 pixels (the
number of rows), and a depth of 3 (the number of channels). To access an individual pixel value
from our **image** we use simple NumPy array indexing:

In [ ]:
(r, g, b) = image[20, 100]) # accesses pixel at x=100, y=20

OpenCV stores **RGB** channel pixels value in reverse order as **BGR**.

## Image Classification Basics

### Step 1 - Gather datset
We need images with labels associated with them. The number of images in each category should be uniform. If we increase the number of images in any categories then classifier will became naturally baised to overfitting into these heavily represented categories. Avoid *class imbalance* completely.

### Step 2 - Split your dataset
Splitting dataset into training set and test set. It’s extremely important that the training set and testing set are independent of each other and do not overlap! We split dataset in the following ratio. <img src="files/4.png">
In practice, we need to test a bunch of these hyperparameters and identify the set of parameters that works the best. **We normally allocate roughly 10-20% of the training data for validation.**

### Step 3 - Training a network
This is what deep learning is about.

### Step 4 - Evaluate